### PR TITLE
ISPN-9366 Cache.size optimizations (pt. 2)

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-9366

* Created a method for the size in the Publisher* interface and classes.

* We query locally for the size and issue request to some of the remote
targets to do the same. With the returned value we sum everything.

* Change ComponentRegistry to create the LocalClusterPublisherManager,
only needed so the component is already started when used.

This relates to #10042

### CODE REVIEW

File: core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java [hunk 1]
```diff
@@ -861,6 +920,13 @@ public <R> SegmentPublisherSupplier<R> entryPublisher(IntSet segments, Set<K> ke
                                              deliveryGuarantee, batchSize, transformer);
    }
 
+   @Override
+   public CompletionStage<Long> sizePublisher(IntSet segments, Set<K> exclude, InvocationContext ctx, long flags) {
+      long addedFlags = EnumUtil.mergeBitSets(flags, FlagBitSets.CACHE_MODE_LOCAL);
+      return reduction(false, segments, null, ctx, addedFlags, DeliveryGuarantee.EXACTLY_ONCE,
+            sizeComposedType(), s -> PublisherReducers.count().apply(s), s -> PublisherReducers.add().apply(s));
```
Comment (wburns) [inline core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:927]: IIUC the transformer is not used in this case. I would suggest just setting it to null to confirm or even throwing an exception.

File: core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java [hunk 2]
```diff
@@ -825,6 +840,50 @@ public CacheEntry<K, V> fromCacheEntry(CacheEntry entry) {
...
+                                          DeliveryGuarantee deliveryGuarantee,
+                                          Function<? super Publisher<Long>, ? extends CompletionStage<Long>> transformer,
+                                          Function<? super Publisher<Long>, ? extends CompletionStage<Long>> finalizer) {
+         return commandsFactory.buildSizeCommand(segments, keysToExclude, explicitFlags);
+      }
+
+      @Override
+      public CompletionStage<PublisherResult<Long>> contextInvocation(IntSet segments, Set<K> keysToInclude,
+         InvocationContext ctx, Function<? super Publisher<Long>, ? extends CompletionStage<Long>> transformer) {
+         return KEY_COMPOSED.contextInvocation(segments, keysToInclude, ctx, transformer);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:868]: This shouldn't ever be used. If the user provides a context that is not empty we either have to skip the optimization or we could do the optimization for segments that aren't in the context. I would say for this PR just always disable the optimization if the context has any values.

File: core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java [hunk 3]
```diff
@@ -825,6 +840,50 @@ public CacheEntry<K, V> fromCacheEntry(CacheEntry entry) {
...
 
+   private class SizeComposedType implements ComposedType<K, Long, Long> {
+
+      @Override
+      public CompletionStage<PublisherResult<Long>> localInvocation(boolean parallelPublisher,
+         IntSet segments, Set<K> keysToInclude, Set<K> keysToExclude, long explicitFlags,
+         DeliveryGuarantee deliveryGuarantee,
+         Function<? super Publisher<Long>, ? extends CompletionStage<Long>> transformer,
+         Function<? super Publisher<Long>, ? extends CompletionStage<Long>> finalizer) {
+         return KEY_COMPOSED.localInvocation(parallelPublisher, segments, keysToInclude, keysToExclude,
```
Comment (wburns) [inline core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java:851]: This should use the SizeCommand as well I would expect.

**Review by wburns — CHANGES_REQUESTED** [review]
I like the use of the composed type 👍

File: core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java [hunk 4]
```diff
@@ -145,6 +154,21 @@ protected Boolean addBooleanResponse(Address sender, Boolean response) {
...
+         if (command.hasAnyFlag(FlagBitSets.SKIP_SIZE_OPTIMIZATION) || !usingStores()) {
+            return asyncValue(clusterPublisherManager.keyReduction(false, null, null, ctx,
+                  command.getFlagsBitSet(), DeliveryGuarantee.EXACTLY_ONCE, PublisherReducers.count(),
+                  PublisherReducers.add()));
+         }
+
+         return asyncValue(clusterPublisherManager.sizePublisher(command.getSegments(), ctx, command.getFlagsBitSet()));
+      }
+
+      return invokeNext(ctx, command);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java:169]: I would expect this to look more like this. The `sizePublisher` would then pass the FLAG_MODE_LOCAL flag.

```suggestion
      if (isLocalModeForced(command) {
         return invokeNext(ctx, command);
      }
      if (command.hasAnyFlag(FlagBitSets.SKIP_SIZE_OPTIMIZATION)) {
         return asyncValue(clusterPublisherManager.keyReduction(false, null, null, ctx,
               command.getFlagsBitSet(), DeliveryGuarantee.EXACTLY_ONCE, PublisherReducers.count(),
               PublisherReducers.add()));
      }

      return asyncValue(clusterPublisherManager.sizePublisher(command.getSegments(), ctx, command.getFlagsBitSet()));
```

File: core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java [hunk 5]
```diff
@@ -385,6 +387,16 @@ public void segmentsLost(IntSet lostSegments) {
       changeListener.forEach(lostSegments::forEach);
    }
 
+   @Override
+   public CompletionStage<Long> sizePublisher(IntSet segments, long flags) {
+      if (!EnumUtil.containsAny(flags, FlagBitSets.CACHE_MODE_LOCAL)) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java:392]: This if shouldn't be needed afai can tell.

File: core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java (hunk 4 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java:169]: I am not sure why the usingStores was referenced. We should be able to do the optimization with or without stores.

File: core/src/main/java/org/infinispan/container/impl/InternalDataContainer.java [hunk 6]
```diff
@@ -139,7 +139,7 @@ default int size(IntSet segments) {
       int size = 0;
       // We have to loop through and count the entries
       for (Iterator<InternalCacheEntry<K, V>> iter = iterator(segments); iter.hasNext(); ) {
-         iter.next();
+         if (iter.next().getValue() == null) continue;
```
Comment (jabolina) [inline core/src/main/java/org/infinispan/container/impl/InternalDataContainer.java:142]: because of the scattered mode

**Comment by jabolina** [comment]
Moved everything to #10042

**Pull request closed without merging**